pkg/queryfrontend/cache/memcached.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package cache
+
+import (
+	"context"
+	"time"
+
+	cortexcache "github.com/cortexproject/cortex/pkg/chunk/cache"
+	"github.com/cortexproject/cortex/pkg/querier/queryrange"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"gopkg.in/yaml.v2"
+
+	"github.com/thanos-io/thanos/pkg/cacheutil"
+)
+
+const (
+	memcachedDefaultTTL = 24 * time.Hour
+)
+
+// MemcachedCache is a memcached-based cache.
+type MemcachedCache struct {
+	logger    log.Logger
+	memcached cacheutil.MemcachedClient
+	validity  time.Duration
+
+	// Metrics.
+	requests prometheus.Counter
+	hits     prometheus.Counter
+}
+
+type MemcachedResponseCacheConfig struct {
+	Memcached cacheutil.MemcachedClientConfig `yaml:",inline"`
+	Validity  time.Duration                   `yaml:"validity"`
+}
+
+// newMemcachedCache makes a new MemcachedCache and returns Cortex ResultsCacheConfig.
+func newMemcachedCache(conf []byte, logger log.Logger, reg prometheus.Registerer) (*queryrange.ResultsCacheConfig, error) {
+	var config MemcachedResponseCacheConfig
+	if err := yaml.UnmarshalStrict(conf, &config); err != nil {
+		return nil, err
+	}
+
+	memcachedClientConfig, err := yaml.Marshal(config.Memcached)
+	if err != nil {
+		return nil, errors.Wrap(err, "marshal content of memcached client configuration")
+	}
+
+	memcached, err := cacheutil.NewMemcachedClient(logger, "response-cache", memcachedClientConfig, reg)
+	if err != nil {
+		return nil, errors.Wrap(err, "create memcached client")
+	}
+
+	c := newMemcachedCacheWithClient(memcached, config.Validity, logger, reg)
+
+	return &queryrange.ResultsCacheConfig{
+		CacheConfig: cortexcache.Config{Cache: c},
+	}, nil
+}
+
+func newMemcachedCacheWithClient(
+	memcached cacheutil.MemcachedClient,
+	validity time.Duration,
+	logger log.Logger,
+	reg prometheus.Registerer,
+) *MemcachedCache {
+	c := &MemcachedCache{
+		logger:    logger,
+		memcached: memcached,
+		validity:  validity,
+	}
+
+	if c.validity == 0 {
+		level.Info(logger).Log("msg", "memcached cache valid time set to 0, use 24 hours instead")
+		c.validity = memcachedDefaultTTL
+	}
+
+	c.requests = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name: "thanos_cache_memcached_requests_total",
+		Help: "Total number of items requests to memcached.",
```

Comment:
Is this really `items requests`? Or could it be `Total number of requests to memcached.`

Reply:
This metric is for `items requests`, same as https://github.com/thanos-io/thanos/blob/master/pkg/cache/memcached.go#L35.

Yes, it is not very clear about the naming.

Reply:
yes, I suspected this was the case. It reads a little bit funny, like we changed our mind while writing the sentence, but not too serious.